wallet/wallet-controller/src/read.rs

```diff
@@ -179,30 +194,75 @@ where
             .map_err(ControllerError::WalletError)
     }
 
-    pub fn get_addresses_usage(&self) -> Result<&'a KeychainUsageState, ControllerError<T>> {
+    pub fn get_addresses_usage(
+        &self,
+        key_purpose: KeyPurpose,
+    ) -> Result<&'a KeychainUsageState, ControllerError<T>> {
         self.wallet
-            .get_addresses_usage(self.account_index)
+            .get_addresses_usage(self.account_index, key_purpose)
             .map_err(ControllerError::WalletError)
     }
 
-    /// Get all addresses with usage information
-    /// The boolean in the BTreeMap's value is true if the address is used, false is otherwise
+    /// Get all addresses with usage information and coin balances
+    /// The used boolean field is true if the address is used, false is otherwise.
     /// Note that the usage statistics follow strictly the rules of the wallet. For example,
     /// the initial wallet only stored information about the last used address, so the usage
     /// of all addresses after the first unused address will have the result `false`.
     pub fn get_addresses_with_usage(
         &self,
-    ) -> Result<MapAddressWithUsage<Destination>, ControllerError<T>> {
-        let addresses = self.get_all_issued_addresses()?;
-        let usage = self.get_addresses_usage()?;
-
-        Ok(addresses
-            .into_iter()
-            .map(|(child_number, address)| {
+        include_change_addresses: bool,
+    ) -> Result<Vec<AddressInfo>, ControllerError<T>> {
+        fn map_addresses<'a, I>(
```

Comment:
why fn and not a closure?

Reply:
closures can't have generic inputs which is the 2 different iterators one with a Chain and one without.